File: src/wih/github_client.py
from __future__ import annotations

import json
import urllib.parse
import urllib.request


class GitHubClient:
    def __init__(self, repo: str, token: str) -> None:
        if "/" not in repo:
            raise ValueError("repo must be in owner/name format")
        if not token:
            raise ValueError("missing GitHub token")
        self.repo = repo
        self.token = token

    def list_issues(self, state: str = "all") -> list[dict]:
        owner, name = self.repo.split("/", 1)
        per_page = 100
        page = 1
        rows: list[dict] = []

        while True:
            query = urllib.parse.urlencode({"state": state, "per_page": per_page, "page": page})
            url = f"https://api.github.com/repos/{owner}/{name}/issues?{query}"
            request = urllib.request.Request(
                url,
                headers={
                    "Authorization": f"Bearer {self.token}",
                    "Accept": "application/vnd.github+json",
                    "X-GitHub-Api-Version": "2022-11-28",
                    "User-Agent": "who-is-hiring-builder",
                },
            )
            with urllib.request.urlopen(request, timeout=30) as response:
                payload = json.loads(response.read().decode("utf-8"))

            if not isinstance(payload, list):
                raise RuntimeError("unexpected GitHub API response")

            issues_only = [item for item in payload if "pull_request" not in item]
            rows.extend(issues_only)

            if len(payload) < per_page:
                break
            page += 1

        return rows

File: src/githubClient.ts
import { githubIssueSchema, type GitHubIssue } from "./schemas.js";

type LabelPayload = {
  name: string;
};

export type GitHubIssueComment = {
  body: string | null;
  created_at: string;
  user: {
    type: string;
  };
};

export class GitHubClient {
  private readonly owner: string;
  private readonly name: string;

  constructor(
    private readonly repo: string,
    private readonly token: string,
  ) {
    if (!repo.includes("/")) {
      throw new Error("repo must be in owner/name format");
    }
    if (!token) {
      throw new Error("missing GitHub token");
    }

    [this.owner, this.name] = repo.split("/", 2);
  }

  async listIssues(state = "all"): Promise<GitHubIssue[]> {
    const perPage = 100;
    let page = 1;
    const rows: GitHubIssue[] = [];

    while (true) {
      const params = new URLSearchParams({
        state,
        per_page: String(perPage),
        page: String(page),
      });
      const payload = await this.requestJson<unknown[]>(`/issues?${params.toString()}`);

      if (!Array.isArray(payload)) {
        throw new Error("unexpected GitHub API response");
      }

      const issuesOnly = payload.filter((item) => !(item as { pull_request?: unknown }).pull_request);
      const parsedIssues = issuesOnly.map((item) => githubIssueSchema.parse(item));
      rows.push(...parsedIssues);

      if (payload.length < perPage) {
        break;
      }
      page += 1;
    }

    return rows;
  }

  async ensureLabelExists(name: string, color = "d4c5f9", description = "Issue is missing key hiring details"): Promise<void> {
    const labels = await this.requestJson<LabelPayload[]>("/labels?per_page=100");
    const exists = labels.some((label) => label.name.toLowerCase() === name.toLowerCase());
    if (exists) {
      return;
    }

    await this.requestJson("/labels", {
      method: "POST",
      body: JSON.stringify({ name, color, description }),
    });
  }

  async addLabelToIssue(issueNumber: number, label: string): Promise<void> {
    await this.requestJson(`/issues/${issueNumber}/labels`, {
      method: "POST",
      body: JSON.stringify({ labels: [label] }),
    });
  }

  async listIssueComments(issueNumber: number): Promise<GitHubIssueComment[]> {
    return this.requestJson<GitHubIssueComment[]>(`/issues/${issueNumber}/comments?per_page=100`);
  }

  async createIssueComment(issueNumber: number, body: string): Promise<void> {
    await this.requestJson(`/issues/${issueNumber}/comments`, {
      method: "POST",
      body: JSON.stringify({ body }),
    });
  }

  private async requestJson<T>(path: string, init?: RequestInit): Promise<T> {
    const response = await fetch(`https://api.github.com/repos/${this.owner}/${this.name}${path}`, {
      ...init,
      headers: {
        Authorization: `Bearer ${this.token}`,
        Accept: "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
        "User-Agent": "who-is-hiring-builder",
        ...(init?.headers ?? {}),
      },
    });

    if (!response.ok) {
      throw new Error(`GitHub API failed (${response.status})`);
    }

    return (await response.json()) as T;
  }
}
